### PR TITLE
okta custom roles

### PR DESCRIPTION
**Pull request by mchavez**
- Adding tests and refactoring code
- Fix issue when listing entitlements
- Listing custom roles

### CODE REVIEW

File: pkg/connector/role.go [hunk 1]
```diff
@@ -183,15 +222,72 @@ func (o *roleResourceType) listSystemRoles(
...
+		}
+
+		rv = append(rv, resource)
+	}
+
+	return rv, nil
+}
+
+func listOktaIamCustomRoles(ctx context.Context, client *okta.Client, token *pagination.Token, qp *query.Params) ([]*okta.Role, *responseContext, error) {
+	url := "/api/v1/iam/roles"
```
Comment (mgaeta) [inline pkg/connector/role.go:255]: This should be a constant

Comment (mgaeta) [reply pkg/connector/role.go:255]: There is a lot of repeated code here. I would do something like 
```go
const (
	apiPathListAdministrators = "/api/internal/administrators"
	apiPathListIamCustomRoles = "/api/v1/iam/roles"
	ContentType               = "application/json"
)

func doRequest(
	ctx context.Context,
	client *okta.Client,
	path string,
	queryString string,
	target interface{},
) (*okta.Response, error) {
	reqUrl, err := url.Parse(path)
	if err != nil {
		return nil, err
	}

	if queryString != "" {
		reqUrl.RawQuery = queryString
	}

	rq := client.CloneRequestExecutor()

	req, err := rq.
		WithAccept(ContentType).
		WithContentType(ContentType).
		NewRequest(http.MethodGet, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	
	return rq.Do(ctx, req, &target)
}
```

**Review by mgaeta — APPROVED** [review]

File: pkg/connector/role.go (hunk 1 above)
Comment (mchavez) [reply pkg/connector/role.go]: Fixed

File: pkg/connector/role.go [hunk 2]
```diff
@@ -84,19 +129,18 @@ func (o *roleResourceType) Entitlements(
...
+		en := sdkEntitlement.NewAssignmentEntitlement(resource, "assigned",
+			sdkEntitlement.WithDisplayName(fmt.Sprintf("%s Role Member", role.Label)),
+			sdkEntitlement.WithDescription(fmt.Sprintf("Has the %s role in Okta", role.Label)),
+			sdkEntitlement.WithAnnotation(&v2.V1Identifier{
+				Id: V1MembershipEntitlementID(role.Type),
+			}),
+			sdkEntitlement.WithGrantableTo(resourceTypeUser),
+		)
+		rv = append(rv, en)
+	}
```
Comment (ggreer) [inline pkg/connector/role.go:143]: This skips syncing entitlements for custom roles, and therefore skips syncing grants for them.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -61,24 +62,19 @@ jobs:
...
       - name: Check for grant before revoking
         run: |
-          ./baton-okta --api-token ${{ env.BATON_API_TOKEN }} --domain ${{ env.BATON_DOMAIN }}
-          baton grants --entitlement ${{ env.CONNECTOR_ENTITLEMENT }} --output-format=json | jq -e ".grants | any(.principal.id.resource ==\"${{ env.CONNECTOR_PRINCIPAL }}\")"
+          ./baton-okta && baton grants --entitlement ${{ env.CONNECTOR_ENTITLEMENT }} --output-format=json | jq -e ".grants | any(.principal.id.resource ==\"${{ env.CONNECTOR_PRINCIPAL }}\")"
       - name: Revoke grants
         run: |
-          ./baton-okta --api-token ${{ env.BATON_API_TOKEN }} --domain ${{ env.BATON_DOMAIN }}
-          ./baton-okta --api-token ${{ env.BATON_API_TOKEN }} --domain ${{ env.BATON_DOMAIN }} --revoke-grant ${{ env.CONNECTOR_GRANT }}
+          ./baton-okta && ./baton-okta --revoke-grant ${{ env.CONNECTOR_GRANT }}
```
[the remark this answers is not in the document]
Comment (mchavez) [reply .github/workflows/ci.yaml:71]: Fixed

File: pkg/connector/role.go [hunk 4]
```diff
@@ -40,42 +41,86 @@ var standardRoleTypes = []*okta.Role{
...
 		return nil, "", nil, fmt.Errorf("okta-connectorv2: failed to parse page token: %w", err)
 	}
 
+	if bag.Current() == nil {
+		// Push onto stack in reverse
+		bag.Push(pagination.PageState{
+			ResourceTypeID: resourceTypeRole.Id,
+			ResourceID:     listRoleStandard,
+		})
+		bag.Push(pagination.PageState{
```
[the remark this answers is not in the document]
Comment (mchavez) [reply pkg/connector/role.go:89]: Fixed

File: pkg/connector/user.go [hunk 5]
```diff
@@ -170,6 +170,15 @@ func listUsers(ctx context.Context, client *okta.Client, token *pagination.Token
 	return oktaUsers, respCtx, nil
 }
 
+func listAllUsers(ctx context.Context, client *okta.Client, qp *query.Params) ([]*okta.User, error) {
+	oktaUsers, _, err := client.User.ListUsers(ctx, qp)
```
Comment (ggreer) [inline pkg/connector/user.go:174]: ListUsers is paginated and only returns up to 200 results at a time, so this will only fetch the first 200 users.

**Review by ggreer — APPROVED** [review]

**Pull request merged**